File: lexio/lib/utils/answer-verification/tokenizer.py
import sys
import tiktoken
import json

enc = tiktoken.get_encoding("cl100k_base")

def count_tokens(sentences):
    return [len(enc.encode(sentence)) for sentence in sentences]

if __name__ == "__main__":
    try:
        text = sys.stdin.read().strip()
        sentences = json.loads(text)  # Expecting a JSON list of sentences
        print(json.dumps(count_tokens(sentences)))  # Return token counts as JSON
    except Exception as e:
        print(json.dumps([]))  # Return empty list if an error occurs

File: lexio/lib/explanation/tokenizer.ts
// tokenizer.ts
import { loadTiktoken } from './dependencies';

let encoder: any = null;

/**
 * Initialize the encoder lazily.
 */
async function initializeEncoder() {
  if (!encoder) {
    const tiktoken = await loadTiktoken();
    encoder = tiktoken.getEncoding("cl100k_base");
  }
  return encoder;
}

/**
 * Count tokens for an array of sentence strings.
 *
 * @param sentences - An array of sentences.
 * @returns A promise that resolves to an array of token counts corresponding to each sentence.
 */
export async function countTokens(sentences: string[]): Promise<number[]> {
  const enc = await initializeEncoder();
  return sentences.map((sentence) => enc.encode(sentence).length);
}
